crate/server/src/core/operations/validate.rs

```diff
@@ -0,0 +1,586 @@
+    let (certificate_validity_start, certificate_validity_end) =
+        (certificate.not_before(), certificate.not_after());
+    if date.as_ref() <= certificate_validity_end && certificate_validity_start <= date.as_ref() {
+        println!("DATE IS VALID");
+        KResult::Ok(ValidityIndicator::Valid)
+    } else {
+        println!("DATE IS INVALID");
+        KResult::Ok(ValidityIndicator::Invalid)
+    }
+}
+
+// getting crl uri for all the chain.
+// returns a vector
+fn get_crl_uris_from_certificate_chain(
+    certificates: &mut [Vec<u8>],
+) -> KResult<Vec<(String, Vec<u8>)>> {
+    let init = &mut Vec::<(String, Vec<u8>)>::new();
+    let res = certificates.iter().try_fold(init, |acc, certificate| {
+        let res = &mut get_crl_uri_from_certificate(certificate)?;
+        acc.append(res);
+        KResult::Ok(acc)
+    })?;
+    KResult::Ok(res.clone())
+}
+
+fn get_crl_uri_from_certificate(certificate: &[u8]) -> KResult<Vec<(String, Vec<u8>)>> {
+    /* and vec<u8> */
+    let certificate = X509::from_der(certificate)?;
+    let certificate_hash = certificate.authority_key_id();
+    let certificate_hash = if let Some(auth_id) = certificate_hash {
+        auth_id.as_slice().to_vec()
+    } else {
+        HEAD.to_vec()
+    };
+    let crl_dp = certificate.crl_distribution_points();
+    match crl_dp {
+        None => KResult::Ok([].to_vec()),
+        Some(crl_dp) => {
+            let crl_size = crl_dp.len();
+            let uri_list = &mut Vec::<String>::new();
+            for i in 0..crl_size {
+                let crl_uri = crl_dp
+                    .get(i)
+                    .and_then(DistPointRef::distpoint)
+                    .and_then(DistPointNameRef::fullname)
+                    .and_then(|x| x.get(0))
+                    .and_then(GeneralNameRef::uri);
+                if let Some(crl_uri) = crl_uri {
+                    uri_list.push(crl_uri.to_string())
+                }
+            }
+            let res = uri_list
+                .iter()
+                .map(|x| (x.clone(), certificate_hash.clone()))
+                .collect();
+            KResult::Ok(res)
+        }
+    }
+}
+
+fn crl_status_to_validity_indicator(status: CrlStatus) -> ValidityIndicator {
+    match status {
+        CrlStatus::NotRevoked => {
+            println!("Valid");
+            ValidityIndicator::Valid
+        }
+        CrlStatus::RemoveFromCrl(_) => {
+            println!("Unknown");
+            ValidityIndicator::Invalid
+        }
+        CrlStatus::Revoked(_) => {
+            println!("Invalid");
+            ValidityIndicator::Invalid
+        }
+    }
+}
+
+fn chain_revocation_status(
+    certificates: &[Vec<u8>],
+    crls: &mut [Vec<u8>],
+) -> KResult<ValidityIndicator> {
+    certificates
+        .iter()
+        .try_fold(ValidityIndicator::Valid, |acc, certificate| {
+            let res = certificate_revocation_status(certificate, crls)?;
+            KResult::Ok(acc.and(res))
+        })
```

Comment:
Maybe prefer separation of concerns:
```suggestion
    certificates
        .iter()
        .map(|cert| certificate_revocation_status(cert, crls))
        .try_fold(ValidityIndicator::Valid, |s1, s2| Ok(s1.and(s2?)))
```